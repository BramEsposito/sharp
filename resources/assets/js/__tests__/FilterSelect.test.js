import FilterSelect from '../components/list/FilterSelect.vue';
import Select from '../components/form/fields/Select.vue';
import { shallowMount } from '@vue/test-utils';
import { MockI18n } from "./utils";
import Vue from 'vue';


describe('filter-select', ()=>{
    Vue.use(MockI18n, { mockFn:true });

    function createWrapper(options={}) {
        return shallowMount(FilterSelect, {
            ...options,
            propsData : {
                filterKey: 'job',
                name:'Web job',
                values: [
                    { id:1, label:'front' },
                    { id:2, label:'back' },
                ],
                ...options.propsData
            },
        });
    }

    function findSelect(wrapper) {
        return wrapper.find(Select);
    }

    test('can mount "empty" FilterSelect', ()=>{
        expect(createWrapper().html()).toMatchSnapshot();
    });

    test('can mount "valuated" FilterSelect', ()=>{
        expect(createWrapper({propsData: { value:1 }}).html()).toMatchSnapshot();
    });

    test('can mount "multiple empty" FilterSelect', ()=>{
        expect(createWrapper({ propsData:{multiple:true} }).html()).toMatchSnapshot();
    });

    test('can mount "multiple valuated" FilterSelect', ()=>{
        expect(createWrapper({ propsData:{multiple:true, value: [1,2]} }).html()).toMatchSnapshot();
    });

    test('has select', ()=>{
        expect(findSelect(createWrapper()).isVueInstance()).toBe(true);
    });

    test('expose appropriate props to select', ()=>{
        const wrapper = createWrapper({
            propsData: {
                multiple: true,
                required: true,
                value: [1,2]
            }
        });
        const select = findSelect(wrapper);

        expect(select.vm.$props).toMatchObject({
            value: [1,2],
            options: [
                { id:1, label:'front' },
                { id:2, label:'back' },
            ],
            multiple: true,
            clearable: false,
            inline: false,
            uniqueIdentifier: 'job'
        });
    });

    test('call appropriate handlers', async ()=>{
        const wrapper = createWrapper({
            created() {
                this.handleSelect = jest.fn();
            }
        });
        const label = wrapper.find('.SharpFilterSelect__text');
        const select = findSelect(wrapper);

        wrapper.setMethods({
            showDropdown: jest.fn()
        });

        select.vm.$emit('input');
        expect(wrapper.vm.handleSelect).toHaveBeenCalled();
        label.trigger('mousedown');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.showDropdown).toHaveBeenCalled();
    });

    test('empty', ()=>{
        const wrapper = createWrapper();

        expect(wrapper.vm.empty).toBe(true);
        wrapper.setProps({ value:1 });
        expect(wrapper.vm.empty).toBe(false);
        wrapper.setProps({ value:[], multiple:true });
        expect(wrapper.vm.empty).toBe(true);
        wrapper.setProps({ value:[1], multiple:true });
        expect(wrapper.vm.empty).toBe(false);
    });

    test('emit input', ()=>{
        const wrapper = createWrapper();

        wrapper.vm.handleSelect(1);
        expect(wrapper.emitted('input')).toEqual([[1]]);
    });
});