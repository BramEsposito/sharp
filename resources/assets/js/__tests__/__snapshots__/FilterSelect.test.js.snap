// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter-select can mount "empty" FilterSelect 1`] = `
<span tabindex="0" class="SharpFilterSelect SharpFilterSelect--empty">
	<sharp-autocomplete-stub mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1" searchkeys="value" noresultitem="true"
	    allowempty="true" class="SharpFilterSelect__select"></sharp-autocomplete-stub>
	<span class="SharpFilterSelect__text">
		Web job
	</span>
	<sharp-select-stub uniqueidentifier="job" options="[object Object],[object Object]" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></sharp-select-stub>
</span>
`;

exports[`filter-select can mount "multiple empty" FilterSelect 1`] = `
<span tabindex="0" class="SharpFilterSelect SharpFilterSelect--empty SharpFilterSelect--multiple">
	<sharp-autocomplete-stub value="" mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1" searchkeys="value" noresultitem="true"
	    multiple="true" hideselected="true" allowempty="true" class="SharpFilterSelect__select"></sharp-autocomplete-stub>
	<span class="SharpFilterSelect__text">
		Web job
	</span>
	<sharp-select-stub uniqueidentifier="job" options="[object Object],[object Object]" multiple="true" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></sharp-select-stub>
</span>
`;

exports[`filter-select can mount "multiple valuated" FilterSelect 1`] = `
<span tabindex="0" class="SharpFilterSelect SharpFilterSelect--multiple">
	<sharp-autocomplete-stub value="[object Object],[object Object]" mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1"
	    searchkeys="value" noresultitem="true" multiple="true" hideselected="true" allowempty="true" class="SharpFilterSelect__select"></sharp-autocomplete-stub>
	<span class="SharpFilterSelect__text">
		Web job
	</span>
	<sharp-select-stub value="1,2" uniqueidentifier="job" options="[object Object],[object Object]" multiple="true" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></sharp-select-stub>
</span>
`;

exports[`filter-select can mount "valuated" FilterSelect 1`] = `
<span tabindex="0" class="SharpFilterSelect">
	<sharp-autocomplete-stub value="[object Object]" mode="local" localvalues="[object Object],[object Object]" placeholder="{{ entity_list.filter.search_placeholder }}" remotesearchattribute="query" itemidattribute="id" searchminchars="1" searchkeys="value"
	    noresultitem="true" allowempty="true" class="SharpFilterSelect__select"></sharp-autocomplete-stub>
	<span class="SharpFilterSelect__text">
		Web job
	</span>
	<sharp-select-stub value="1" uniqueidentifier="job" options="[object Object],[object Object]" display="dropdown" clearable="true" placeholder=" " class="SharpFilterSelect__select"></sharp-select-stub>
</span>
`;
